File: Demo/JCTiledViewDemo/RootViewController.m
//
//  RootViewController.m
//  JCTiledViewDemo
//
//  Created by Jesse Collis
//  Copyright (c) 2012 JC Multimedia Design. All rights reserved.
//

#import "RootViewController.h"
#import "JCTiledPDFScrollView.h"
#import "DemoAnnotationView.h"
#import "DemoAnnotation.h"
#import "math.h"
#import "DetailView.h"

#define SkippingGirlImageSize CGSizeMake(432., 648.)

#ifdef ANNOTATE_TILES
#import "JCTiledView.h"
#endif

@interface RootViewController ()

@property (nonatomic, strong) NSArray *bezierPaths;
@end

@implementation RootViewController


#pragma mark - View lifecycle

- (void)loadView
{
  UIView *view = [[UIView alloc] initWithFrame:CGRectMake(0.0f, 0.0f, 320.0f, 460.0f)];
  view.autoresizingMask = UIViewAutoresizingFlexibleWidth|UIViewAutoresizingFlexibleHeight;
  self.view = view;
}

- (void)viewDidLoad
{
  [super viewDidLoad];

  self.detailView = [[DetailView alloc] initWithFrame:CGRectZero];
  self.detailView.autoresizingMask = UIViewAutoresizingFlexibleWidth;
  CGSize size_for_detail = [self.detailView sizeThatFits:self.view.bounds.size];
  [self.detailView setFrame:CGRectMake(0.f,0.f,size_for_detail.width,size_for_detail.height)];
  [self.view addSubview:self.detailView];

  CGRect scrollView_frame = CGRectOffset(CGRectInset(self.view.bounds,0.,size_for_detail.height/2.0f),0.,size_for_detail.height/2.0f);

  //PDF
  //self.scrollView = [[JCTiledPDFScrollView alloc] initWithFrame:scrollView_frame URL:[[NSBundle mainBundle] URLForResource:@"Map" withExtension:@"pdf"]];

  //Bitmap
  self.scrollView = [[JCTiledScrollView alloc] initWithFrame:scrollView_frame contentSize:SkippingGirlImageSize];

  self.scrollView.dataSource = self;
  self.scrollView.tiledScrollViewDelegate = self;
  self.scrollView.zoomScale = 1.0f;

#ifdef ANNOTATE_TILES
  self.scrollView.tiledView.shouldAnnotateRect = YES;
#endif
  
  // totals 4 sets of tiles across all devices, retina devices will miss out on the first 1x set
  self.scrollView.levelsOfZoom = 3;
  self.scrollView.levelsOfDetail = 3;

  [self.view addSubview:self.scrollView];
  
  UIButton *addButton = [UIButton buttonWithType:UIButtonTypeRoundedRect];
  [addButton setTitle:@"+ Annotations" forState:UIControlStateNormal];
  addButton.frame = CGRectMake(CGRectGetWidth(self.view.frame) - 115, 25., 110, 38);
  addButton.autoresizingMask = UIViewAutoresizingFlexibleLeftMargin;
  [addButton addTarget:self action:@selector(addRandomAnnotations) forControlEvents:UIControlEventTouchUpInside];
  [self.view addSubview:addButton];
    
  [self tiledScrollViewDidZoom:self.scrollView]; //force the detailView to update the frist time
  [self addRandomAnnotations];

}

- (void)viewDidUnload
{
  _scrollView = nil;
  _detailView = nil;

  [super viewDidUnload];
}

- (void)viewDidAppear:(BOOL)animated
{
  [super viewDidAppear:animated];
  
  [self.scrollView refreshAnnotations];
  
  [self becomeFirstResponder];
}

#pragma mark - Rotation

- (BOOL)shouldAutorotateToInterfaceOrientation:(UIInterfaceOrientation)interfaceOrientation
{
  return YES;
}

#pragma mark - Responder

- (BOOL)canBecomeFirstResponder
{
  return YES;
}

- (void)motionEnded:(UIEventSubtype)motion withEvent:(UIEvent *)event
{
  if (event.type == UIEventTypeMotion && event.subtype == UIEventSubtypeMotionShake)
  {
    [self.scrollView removeAllAnnotations];
  }
}

#pragma mark - Annotations

- (void)addRandomAnnotations
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        srand(42);
    });
    
    CGSize size           = SkippingGirlImageSize;
    
    NSMutableArray *paths = [NSMutableArray array];
    
    for (int i = 0; i < 6/2; i++) {
        
        id<JCAnnotation> firstAnnot    = [[DemoAnnotation alloc] init];
        CGPoint firstContentPosition = CGPointMake((float)(rand() % (int)size.width), (float)(rand() % (int)size.height));
        firstAnnot.contentPosition     = firstContentPosition;
        id<JCAnnotation> secondAnnot    = [[DemoAnnotation alloc] init];
        CGPoint secondContentPosition = CGPointMake((float)(rand() % (int)size.width), (float)(rand() % (int)size.height));
        secondAnnot.contentPosition     = secondContentPosition;
        [self.scrollView addAnnotation:firstAnnot];
        [self.scrollView addAnnotation:secondAnnot];
        
        
        UIBezierPath *path    = [UIBezierPath bezierPath];
        [path moveToPoint:CGPointMake(firstContentPosition.x, firstContentPosition.y + 30.f)]; // 30.f is max height of the annotation image
        [path addLineToPoint:CGPointMake(secondContentPosition.x, secondContentPosition.y + 30.f)]; // 30.f is max height of the annotation image
        path.lineWidth = 3.f;
        [paths addObject:path];
    }
    
    self.bezierPaths = [NSArray arrayWithArray:paths];
    
}

#pragma mark - JCTiledScrollViewDelegate

- (void)tiledScrollViewDidZoom:(JCTiledScrollView *)scrollView
{
  self.detailView.textLabel.text = [NSString stringWithFormat:@"zoomScale: %0.2f", scrollView.zoomScale];
}

- (void)tiledScrollView:(JCTiledScrollView *)scrollView didReceiveSingleTap:(UIGestureRecognizer *)gestureRecognizer
{
  CGPoint tapPoint = [gestureRecognizer locationInView:(UIView *)scrollView.tiledView];

  //tap point on the tiledView does not inlcude the zoomScale applied by the scrollView
  self.detailView.textLabel.text = [NSString stringWithFormat:@"zoomScale: %0.2f, x: %0.0f y: %0.0f", scrollView.zoomScale, tapPoint.x, tapPoint.y];
}

- (JCAnnotationView *)tiledScrollView:(JCTiledScrollView *)scrollView viewForAnnotation:(id<JCAnnotation>)annotation
{
  NSString static *reuseIdentifier = @"JCAnnotationReuseIdentifier";
  DemoAnnotationView *view = (DemoAnnotationView *)[scrollView dequeueReusableAnnotationViewWithReuseIdentifier:reuseIdentifier];

  if (!view)
  {
    view = [[DemoAnnotationView alloc] initWithFrame:CGRectZero annotation:annotation reuseIdentifier:@"Identifier"];
    view.imageView.image = [UIImage imageNamed:@"marker-red.png"];
    [view sizeToFit];
  }

  return view;
}

#pragma mark - JCTileSource

#define SkippingGirlImageName @"SkippingGirl"

- (UIImage *)tiledScrollView:(JCTiledScrollView *)scrollView imageForRow:(NSInteger)row column:(NSInteger)column scale:(NSInteger)scale
{
 return [UIImage imageNamed:[NSString stringWithFormat:@"tiles/%@_%ldx_%ld_%ld.png", SkippingGirlImageName, scale, row, column]];
}


-(UIBezierPath *)tiledScrollViewGetBezierPath:(JCTiledScrollView *)scrollView atIndex:(NSInteger)index{
    return [self.bezierPaths objectAtIndex:index];
}

- (UIColor *)tiledScrollViewGetPathColor:(JCTiledScrollView *)scrollView atIndex:(NSInteger)index {
    //UIColor *color = [UIColor colorWithRed:38.f/255.f green:166.f/255.f blue:91.f/255.f alpha:1.f];
    UIColor *color = [UIColor whiteColor];
    return color;
}

- (NSInteger)tiledScrollViewGetNumberOfPath:(JCTiledScrollView *)scrollView {
    return [self.bezierPaths count];
}


@end

File: Demo/JCTiledViewDemo/RootViewController.swift
//
//  RootViewController.swift
//
//  Copyright (c) 2012-2017 Jesse Collis <jesse@jcmultimedia.com.au>
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.
//

import UIKit
import JCTiledScrollView


class RootViewController: UIViewController, JCTileSource, JCTiledScrollViewDelegate {

  var scrollView: JCTiledScrollView!
  @IBOutlet var statusLabel: UILabel!
  @IBOutlet var detailView: UIView!
  
  let skippingGirlImageSize = CGSize(width:432, height:648)
  
  override var canBecomeFirstResponder: Bool {
    return true
  }
  
  override func viewDidLoad() {
    super.viewDidLoad()
    
    let scrollView = JCTiledScrollView(frame: self.view.frame, contentSize: skippingGirlImageSize)
    scrollView.translatesAutoresizingMaskIntoConstraints = false
    self.scrollView = scrollView
    self.view.insertSubview(scrollView, belowSubview: self.detailView)
    
    let views: [String: UIView] = ["label": self.detailView, "scrollView": scrollView]
    
    self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "|[scrollView]|",
                                                            options: [],
                                                            metrics: nil,
                                                            views: views))
    self.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:[label][scrollView]|",
                                                            options: [],
                                                            metrics: nil,
                                                            views: views))
    scrollView.dataSource = self
    scrollView.tiledScrollViewDelegate = self
    scrollView.zoomScale = 1
    
    scrollView.tiledView.shouldAnnotateRect = true
    //  // totals 4 sets of tiles across all devices, retina devices will miss out on the first 1x set
    scrollView.levelsOfZoom = 3
    scrollView.levelsOfDetail = 3
    
    updateStatusLabel()
    addRandomAnnotations()
  }
  
  override func viewDidAppear(_ animated: Bool) {
    super.viewDidAppear(animated)
    
    scrollView.refreshAnnotations()
    
    self.becomeFirstResponder()
  }
  
  @IBAction func addRandomAnnotations(_ sender: Any? = nil) {
    for _ in 1...5 {
      let annotation = DemoAnnotation()
      annotation.contentPosition = CGPoint(x:randomIntFrom(start: 0, to: Int(skippingGirlImageSize.width)),
                                           y: randomIntFrom(start: 0, to: Int(skippingGirlImageSize.height)))
      scrollView.add(annotation)
    }
  }
  
  func randomIntFrom(start: Int, to end: Int) -> Int {
    var a = start
    var b = end
    // swap to prevent negative integer crashes
    if a > b {
      swap(&a, &b)
    }
    return Int(arc4random_uniform(UInt32(b - a + 1))) + a
  }
  
  //MARK: Shake Gesture
  
  override func motionEnded(_ motion: UIEventSubtype, with event: UIEvent?) {
    guard let event = event else { return }
    
    if (event.type == .motion && event.subtype == .motionShake) {
      scrollView.removeAllAnnotations()
    }
  }

  //MARK: TileSource
  
  func tiledScrollView(_ scrollView: JCTiledScrollView, imageForRow row: Int, column: Int, scale: Int) -> UIImage {
    return UIImage(named: "tiles/SkippingGirl_\(scale)x_\(row)_\(column).png")!
  }
  
  //MARK: Annotations
  
  func tiledScrollViewDidZoom(_ scrollView: JCTiledScrollView) {
    updateStatusLabel()
  }
  
  func tiledScrollView(_ scrollView: JCTiledScrollView, didReceiveSingleTap gestureRecognizer: UIGestureRecognizer) {
    let tapPoint = gestureRecognizer.location(in: scrollView.tiledView)
    
    //tap point on the tiledView does not inlcude the zoomScale applied by the scrollView
    self.statusLabel.text = String(format:"zoomScale: %0.2f, x: %0.0f y: %0.0f", scrollView.zoomScale, tapPoint.x, tapPoint.y)
  }
  
  func tiledScrollView(_ scrollView: JCTiledScrollView, viewFor annotation: JCAnnotation) -> JCAnnotationView {
    if let annotationView = scrollView.dequeueReusableAnnotationView(withReuseIdentifier: "ReuseIdentifier") {
      return annotationView
    }
    
    let annotationView = DemoAnnotationView(frame:.zero, annotation:annotation, reuseIdentifier: "ReuseIdentifier")
    annotationView.imageView?.image = UIImage(named:"marker-red.png")
    annotationView.sizeToFit()
    
    return annotationView
  }
  
  func updateStatusLabel() {
    self.statusLabel.text = String(format:"zoomScale: %0.2f", self.scrollView.zoomScale)
  }
}
